### PR TITLE
Iio unregister

### PR DESCRIPTION
**Pull request by cristiantpop**
iio: Release resources properly. 
"iio_unregister" function should return the interface and further more, the
resources allocated should be released in "iio_*_app_remove" functions.

Signed-off-by: Cristian Pop <cristian.pop@analog.com>

### CODE REVIEW

File: iio/iio_app/iio_axi_dac_app.c [hunk 1]
```diff
@@ -922,14 +922,18 @@ int32_t iio_axi_dac_app_init(struct iio_axi_dac_app_desc **desc,
...
+	struct iio_interface *iio_interface;
+	struct iio_axi_dac *iio_axi_dac_inst;
 
 	if (!desc)
 		return FAILURE;
 
-	status = iio_unregister(desc->iio_axi_dac_inst->dac->name);
-	if(status < 0)
-		return status;
+	iio_interface = iio_unregister(desc->iio_axi_dac_inst->dac->name);
```
Comment (mchindri) [inline iio/iio_app/iio_axi_dac_app.c]: Review the design

**Comment by cristiantpop** [comment]
V2:

- Change interface to "iio_unregister" function. Allocate and free it at the same level.

File: iio/iio_app/iio_axi_dac_app.c [hunk 2]
```diff
@@ -923,14 +929,18 @@ int32_t iio_axi_dac_app_init(struct iio_axi_dac_app_desc **desc,
...
 
-	status = iio_unregister(desc->iio_axi_dac_inst->dac->name);
-	if(status < 0)
-		return status;
-
+	status = iio_unregister(desc->iio_interface);
+	if (status < 0) {
+		iio_axi_dac_delete_device(desc->iio_interface->iio);
+		iio_axi_dac_inst = (struct iio_axi_dac *)desc->iio_interface->dev_instance;
+		free(iio_axi_dac_inst);
```
Comment (mchindri) [inline iio/iio_app/iio_axi_dac_app.c]: you can do it directly: `free(desc->iio_interface->dev_instance)`

**Comment by cristiantpop** [comment]
V3:

-  free(desc->iio_interface->dev_instance) directly;

- return error if something fails when freeing.

**Comment by mchindri** [comment]
LGTMMMMM

**Pull request merged**